File: app/src/main/java/pl/tysia/maggstone/ui/presentation_logic/adapter/DocumentAdapter.java
package pl.tysia.maggstone.ui.presentation_logic.adapter;


import android.content.Context;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.EditText;
import android.widget.ImageButton;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.constraintlayout.widget.ConstraintLayout;
import androidx.recyclerview.widget.RecyclerView;

import java.util.ArrayList;

import pl.tysia.maggstone.R;
import pl.tysia.maggstone.data.model.DocumentItem;


public class DocumentAdapter<T extends DocumentItem> extends CatalogAdapter<T, DocumentAdapter<T>.DocumentViewHolder> {

    public DocumentAdapter(ArrayList<T> items) {
        super(items);

    }

    public class DocumentViewHolder extends RecyclerView.ViewHolder implements TextWatcher, View.OnClickListener {
        TextView title;
        TextView description;
        TextView name;
        ConstraintLayout back;
        ImageButton deleteButton;
        ImageButton infoButton;

        ImageView moreButton;
        ImageView lessButton;
        EditText numberET;

        ImageView checkImage;


        DocumentViewHolder(View v) {
            super(v);

            title = v.findViewById(R.id.title_tv);
            description = v.findViewById(R.id.description_tv);
            name = v.findViewById(R.id.name_tv);
            back = v.findViewById(R.id.back);
            deleteButton = v.findViewById(R.id.delete_button);
            infoButton = v.findViewById(R.id.info_button);
            moreButton = v.findViewById(R.id.more_button);
            lessButton = v.findViewById(R.id.less_button);
            numberET = v.findViewById(R.id.number_et);

            checkImage = v.findViewById(R.id.check_image);

            numberET.addTextChangedListener(this);

            moreButton.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View view) {
                    DocumentItem item = shownItems.get(getAdapterPosition());

                    if (item.getIlosc() < Double.MAX_VALUE)
                        item.setIlosc(item.getIlosc() + 1);

                    notifyDataSetChanged();
                }
            });

            lessButton.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View view) {
                    DocumentItem item = (DocumentItem) shownItems.get(getAdapterPosition());

                    if (item.getIlosc() > 0)
                        item.setIlosc(item.getIlosc() - 1);

                    notifyDataSetChanged();

                }
            });

            deleteButton.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View view) {
                    allItems.remove(allItems.get(getAdapterPosition()));
                    filter();
                    notifyDataSetChanged();
                }
            });

            v.setOnClickListener(this);

        }

        @Override
        public void onClick(View v) {
            int pos = getAdapterPosition();
            onItemClick(v, pos);
        }

        @Override
        public void beforeTextChanged(CharSequence charSequence, int i, int i1, int i2) {

        }

        @Override
        public void onTextChanged(CharSequence charSequence, int i, int i1, int i2) {

        }

        @Override
        public void afterTextChanged(Editable editable) {
            DocumentItem item = (DocumentItem) shownItems.get(getAdapterPosition());

            try {
                double quantity = Double.parseDouble(numberET.getText().toString());

                if (quantity >= 0)
                    item.setIlosc(quantity);
                else numberET.setText("0");

            }catch (NumberFormatException ex){
                numberET.setText("0");
            }


        }
    }

    @NonNull
    @Override
    public DocumentViewHolder onCreateViewHolder(@NonNull ViewGroup viewGroup, int i) {
        View v = LayoutInflater.from(viewGroup.getContext())
                .inflate(R.layout.list_item_document, viewGroup, false);

        v.setLayoutParams(new RecyclerView.LayoutParams(RecyclerView.LayoutParams.MATCH_PARENT, RecyclerView.LayoutParams.WRAP_CONTENT));

        return new DocumentViewHolder(v);
    }

    @Override
    public void onBindViewHolder(@NonNull DocumentViewHolder holder, int position) {
        DocumentItem item = shownItems.get(position);
        Context context = holder.back.getContext();

        holder.title.setText(item.getTitle());
        holder.description.setText(item.getDescription());
        holder.name.setText(item.getShortDescription());

        holder.numberET.setText(Double.toString(item.getIlosc()));

    }
}

File: app/src/main/java/pl/tysia/maggstone/ui/presentation_logic/adapter/DocumentAdapter.kt
package pl.tysia.maggstone.ui.presentation_logic.adapter

import android.os.Handler
import android.os.Looper
import android.text.Editable
import android.text.TextWatcher
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.*
import androidx.constraintlayout.widget.ConstraintLayout
import androidx.recyclerview.widget.RecyclerView
import pl.tysia.maggstone.R
import pl.tysia.maggstone.data.model.DocumentItem
import java.lang.Integer.MAX_VALUE
import java.util.*
import kotlin.math.round


open class DocumentAdapter<T : DocumentItem>(items: ArrayList<T>) :
    CatalogAdapter<T, DocumentAdapter<T>.DocumentViewHolder>(items) {

    override var selectedItem: T? = null

    open inner class DocumentViewHolder(v: View) :
        RecyclerView.ViewHolder(v), View.OnClickListener, View.OnFocusChangeListener, TextWatcher {

        var title: TextView = v.findViewById(R.id.title_tv)
        var description: TextView = v.findViewById(R.id.subtitle_tv)
        var name: TextView = v.findViewById(R.id.name_tv)
        var back: ConstraintLayout = v.findViewById(R.id.back)
        var deleteButton: ImageButton = v.findViewById(R.id.delete_button)
        var infoButton: ImageButton = v.findViewById(R.id.info_button)
        var moreButton: ImageView = v.findViewById(R.id.more_button)
        var lessButton: ImageView = v.findViewById(R.id.less_button)
        var numberET: EditText = v.findViewById(R.id.number_et)
        var checkImage: ImageView = v.findViewById(R.id.check_image)

        override fun onClick(v: View) {
            val pos = adapterPosition
            onItemClick(v, pos)
        }

        init {
            numberET.onFocusChangeListener = this
            numberET.addTextChangedListener(this)

            moreButton.setOnClickListener {
                val item: DocumentItem = shownItems[adapterPosition]
                numberET.clearFocus();
                onMoreClicked(item)

                changeListeners.forEach { it.onListChanged() }
            }

            lessButton.setOnClickListener {
                val item = shownItems[adapterPosition] as DocumentItem
                numberET.clearFocus();
                onLessClicked(item)

                changeListeners.forEach { it.onListChanged() }
            }

            deleteButton.setOnClickListener {
               onDelete()
            }

            v.setOnClickListener(this)
        }

        open fun onDelete(){
            allItems.remove(allItems[adapterPosition])
            filter()
            notifyDataSetChanged()

            changeListeners.forEach { it.onListChanged() }
        }

        override fun onFocusChange(v: View?, hasFocus: Boolean) {
            if (!hasFocus) {
                Handler(Looper.getMainLooper()).postDelayed(object : Runnable {
                    override fun run() {
                        numberET.clearFocus()
                        notifyDataSetChanged()
                    }
                },50)
            }
        }

        override fun afterTextChanged(s: Editable?) {
            val item = allItems[adapterPosition]
            if (selectedItem != item)
                selectedItem = item

            if(numberET.text.toString().isEmpty()) {
                item.ilosc=0.0
                fixCount(item,1)
                showError(item, numberET)
            }

            else if (numberET.hasFocus()) {
                try {
                    val quantity = numberET.text.toString().replace(',', '.').toDouble()

                    if (quantity >= 0) item.ilosc = quantity
                    fixCount(item,0)
                } catch (ex: NumberFormatException) {
                    fixCount(item,-1)
                    item.ilosc=0.0
                } finally {
                    //if (numberET.text.toString().replace(',','.').toDouble() != item.ilosc) numberET.setText(countToStr(item.ilosc))


                    showError(item, numberET)
                }

            }
        }

        override fun beforeTextChanged(s: CharSequence?, start: Int, count: Int, after: Int) {
        }

        override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) {
        }
    }

    protected open fun isCountValid(item: T) {
        if(item.ilosc <= 0) item.iloscOk=1;
    }

    /*private fun setCountValidity(item: T){
        if(item.ilosc > 0.0) {
            item.iloscOk = 0
        }

        changeListeners.forEach { it.onListChanged() }
    }*/

    fun fixCount(item: T, err: Int){
        /*if (!isCountValid(item)){
            if (item.ilosc < 0) item.ilosc = 0.0
        }*/
        item.iloscOk=err
        if(item.iloscOk==0) isCountValid(item)

        //setCountValidity(item)
        changeListeners.forEach { it.onListChanged() }
    }

    private fun showError(item: DocumentItem, view: EditText){
        if (item.iloscOk!=0) {
            view.error = getErrorTx(item.iloscOk)
        }
        else{
            view.error = null
        }
    }

    fun getErrorTx(err: Int): String {
        return when (err) {
            -1 -> {
                "Nieprawidłowy format danych"
            }
            1 -> {
                "Ilość musi być większa niż 0"
            }
            2 -> {
                "Zbyt duża ilość"
            }
            else -> ""
        }
    }

    protected open fun onMoreClicked(item: DocumentItem){
        if (item.ilosc < MAX_VALUE) {
            item.ilosc = item.ilosc + 1
            item.iloscOk = 0
        }
        notifyDataSetChanged()
    }

    protected open fun onLessClicked(item: DocumentItem){
        if (item.ilosc > 1) {
            item.ilosc = item.ilosc - 1
            item.iloscOk=0
        }
        notifyDataSetChanged()
    }

    override fun onCreateViewHolder(viewGroup: ViewGroup, i: Int): DocumentViewHolder {
        val v = LayoutInflater.from(viewGroup.context)
            .inflate(R.layout.list_item_document, viewGroup, false)
        v.layoutParams = RecyclerView.LayoutParams(
            RecyclerView.LayoutParams.MATCH_PARENT,
            RecyclerView.LayoutParams.WRAP_CONTENT
        )
        return DocumentViewHolder(v)
    }

    override fun onBindViewHolder(holder: DocumentViewHolder, position: Int) {
        val item: T = shownItems[position]

        holder.title.text = item.getTitle()
        holder.description.text = item.getDescription()
        holder.name.text = item.getSubtitle()
        holder.numberET.setText(countToStr(item.ilosc))

        showError(item, holder.numberET)
    }

    fun countToStr(ilo: Double): String {
        val x = 1.0*ilo.toInt()
        if(x==ilo) return ilo.toInt().toString()
        return ilo.round(4).toString().replace('.', ',')
    }

    fun Double.round(decimals: Int): Double {
        var multiplier = 1.0
        repeat(decimals) { multiplier *= 10 }
        return round(this * multiplier) / multiplier
    }
}